tokio-util/src/sync/poll_semaphore.rs

```diff
@@ -55,10 +53,27 @@ impl PollSemaphore {
     /// the `Waker` from the `Context` passed to the most recent call is
     /// scheduled to receive a wakeup.
     pub fn poll_acquire(&mut self, cx: &mut Context<'_>) -> Poll<Option<OwnedSemaphorePermit>> {
-        let result = ready!(self.permit_fut.poll(cx));
+        let permit_future = match self.permit_fut.as_mut() {
+            Some(fut) => fut,
+            None => {
+                // avoid allocations completely if we can grab a permit immediately
+                match Arc::clone(&self.semaphore).try_acquire_owned() {
+                    Ok(permit) => return Poll::Ready(Some(permit)),
+                    Err(TryAcquireError::Closed) => return Poll::Ready(None),
+                    Err(TryAcquireError::NoPermits) => {}
+                }
+
+                let next_fut = Arc::clone(&self.semaphore).acquire_owned();
+                self.permit_fut = Some(ReusableBoxFuture::new(next_fut));
+
+                self.permit_fut.as_mut().unwrap()
```

Comment:
You can avoid this unwrap with a call to `get_or_insert`.
```suggestion
                self.permit_fut.get_or_insert(ReusableBoxFuture::new(next_fut));
```
Alternatively, the entire match can be replaced with a call to `get_or_insert_with`.

Reply:
Replacing the entire match with a `.get_or_insert` requires cloning the semaphore outside the closure (borrowing issues), which I wanted to avoid having to do. :)

Reply:
Alright, this version is fine then.